common/network-types/src/udp.rs

```diff
@@ -71,15 +71,111 @@ pub enum ForeignDataMode {
+///
+/// The default is `Specific(1)`.
+///
+/// Always use [`into_num_tasks`](UdpStreamParallelism::into_num_tasks) or [`split_evenly_with`](UdpStreamParallelism::split_evenly_with)
+/// to determine the correct number of sockets to spawn.
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub enum UdpStreamParallelism {
+    /// Bind as many sender or receiver sockets as given by [`std::thread::available_parallelism`].
+    Auto,
+    /// Bind specific number of sender or receiver sockets.
+    Specific(NonZeroUsize),
+}
+
+impl Default for UdpStreamParallelism {
+    fn default() -> Self {
+        Self::Specific(NonZeroUsize::MIN)
+    }
+}
+
+impl UdpStreamParallelism {
+    fn avail_parallelism() -> usize {
+        // On non-Linux system this will always default to 1, since the
+        // multiple UDP socket optimization is not possible for those platforms.
+        std::thread::available_parallelism()
+            .map(|n| {
+                if cfg!(target_os = "linux") {
+                    n
+                } else {
+                    NonZeroUsize::MIN
+                }
+            })
+            .unwrap_or_else(|e| {
+                warn!(error = %e, "failed to determine available parallelism, defaulting to 1.");
+                NonZeroUsize::MIN
+            })
+            .into()
+    }
+
+    /// Returns the number of sockets for this and the `other` instance
+    /// when they evenly split the available CPU parallelism.
+    pub fn split_evenly_with(self, other: UdpStreamParallelism) -> (usize, usize) {
+        let cpu_half = (Self::avail_parallelism() / 2).max(1);
+
+        match (self, other) {
+            (UdpStreamParallelism::Auto, UdpStreamParallelism::Auto) => (cpu_half, cpu_half),
+            (UdpStreamParallelism::Specific(a), UdpStreamParallelism::Auto) => {
+                let a = cpu_half.min(a.into());
+                (a, cpu_half * 2 - a)
+            }
+            (UdpStreamParallelism::Auto, UdpStreamParallelism::Specific(b)) => {
+                let b = cpu_half.min(b.into());
+                (cpu_half * 2 - b, b)
+            }
+            (UdpStreamParallelism::Specific(a), UdpStreamParallelism::Specific(b)) => {
+                (cpu_half.min(a.into()), cpu_half.min(b.into()))
+            }
+        }
+    }
+
+    /// Calculates the actual number of tasks for this instance.
+    ///
+    /// The returned value is never more than the maximum available CPU parallelism.
+    pub fn into_num_tasks(self) -> usize {
+        let avail_parallelism = Self::avail_parallelism();
+        match self {
+            UdpStreamParallelism::Auto => avail_parallelism,
+            UdpStreamParallelism::Specific(n) => usize::from(n).min(avail_parallelism),
+        }
+    }
+}
+
+impl From<usize> for UdpStreamParallelism {
+    fn from(value: usize) -> Self {
+        match value {
+            0 => Self::default(),
+            i => Self::Specific(NonZeroUsize::new(i).unwrap()),
+        }
+    }
+}
```

Comment:
TryFrom would be a better choice. This can lead to a runtime panic due to incorrect value passed. Or consider removing the unwrap and replace it with:
```suggestion
impl From<usize> for UdpStreamParallelism {
    fn from(value: usize) -> Self {
        match value {
            0 => Self::default(),
            i => Self::Specific(NonZeroUsize::new(i).unwrap_or(NonZeroUsize::MIN)),
        }
    }
}
```

Reply:
I don't see how it could panic, since 0 is handled.

Reply:
I have unwrap and expect allergy.